logic/atrium-logic-common/src/main/kotlin/ch/tutteli/atrium/logic/creating/iterable/contains/creators/impl/InAnyOrderValuesAssertionCreator.kt

```diff
@@ -38,31 +46,52 @@ class InAnyOrderValuesAssertionCreator<SC, T : IterableLike>(
     override val descriptionNumberOfOccurrences: Translatable = DescriptionIterableAssertion.NUMBER_OF_OCCURRENCES
     override val groupDescription: Translatable = DescriptionIterableAssertion.AN_ELEMENT_WHICH_EQUALS
 
-    override fun getAssertionGroupType(): AssertionGroupType {
-        return if (searchBehaviour is NotSearchBehaviour) {
-            DefaultSummaryAssertionGroupType
-        } else {
-            DefaultListAssertionGroupType
-        }
-    }
-
     override fun makeSubjectMultipleTimesConsumable(container: AssertionContainer<T>): AssertionContainer<List<SC>> =
         turnSubjectToList(container, converter)
 
     override fun search(multiConsumableContainer: AssertionContainer<List<SC>>, searchCriterion: SC): Int =
         multiConsumableContainer.maybeSubject.fold({ -1 }) { subject -> subject.filter { it == searchCriterion }.size }
 
-    override fun decorateAssertion(
-        container: AssertionContainer<List<SC>>,
-        featureAssertion: Assertion
-    ): List<Assertion> {
+    override fun searchAndCreateAssertion(
+        multiConsumableContainer: AssertionContainer<List<SC>>,
+        searchCriterion: SC,
+        featureFactory: (Int, Translatable) -> AssertionGroup
+    ): AssertionGroup {
+        if (searchBehaviour is NotSearchBehaviour) {
+            val list = multiConsumableContainer.maybeSubject.getOrElse { emptyList() }
+            val mismatches = createIndexAssertions(list) { (_, element) -> element == searchCriterion }
+            return if (multiConsumableContainer.hasNext(::identity).holds()) {
+                assertionBuilder.fixedClaimGroup
+                    .withListType
+                    .withClaim(mismatches.isEmpty())
+                    .withDescriptionAndRepresentation(groupDescription, searchCriterion)
+                    .withAssertion(createExplanatoryGroupForMismatches(mismatches))
+                    .build()
+            } else {
+                assertionBuilder.invisibleGroup
+                    .withAssertion(
+                        assertionBuilder.createDescriptive(groupDescription, searchCriterion, falseProvider)
+                    ).build()
```

Comment:
I may be mistaken but I think we this way we don't show to the user the right reason why the expectation failed. For instance, if one writes `expect(setOf<Int>()).notToContain(2)`. Then he would see roughly the following:
```
expected that subject: []        (kotlin.collections.EmptySet <1837188010>)
◆ does not contain:
  ⚬ an element which equals: 2
```
Which is actually a valid statement and should not fail as such (2 is not in the empty list). Thus we should show the following:
```
expected that subject: []        (kotlin.collections.EmptySet <1837188010>)
◆ has next: true
  » does not contain:
    ⚬ an element which equals: 2         (kotlin.Int <1234789>)
```
This way it is clear that `notToContain` is actually `toHaveElementsAndNone`


*Edit*:  We should really refactor this class, the decorating approach isn't any better than the overriding approach laughing:. I see now that we add the hasNext assertion in decorateInAnyOrderAssertion. It's kind of a hack that you use a failing descriptive assertion here (because this part holds) and I therefore suggest that you write a comment here so that the next person sees it immediately. 
```suggestion
                assertionBuilder.invisibleGroup
                    .withAssertion(
                        // TODO 0.18.0 this is a hack so that we see the searchCriterion in the reporting. Refactor so that it is an explanatoryAssertion instead
                        assertionBuilder.createDescriptive(groupDescription, searchCriterion, falseProvider)
                    ).build()
```
Or even better, remove the `if` here and place it in decorateInAnyOrderAssertion where you can turn the `inAnyOrderAssertion` into an explanatory assertion group

Reply:
I took another stab at this, took out the hacks, and simplified the logic. I think earlier I wasn't understanding how to use the explanatoryGroup.

I agree that a refactor may be the best thing for `ContainsObjectsAssertionCreator` and its children. For now it seems like we need both the `decorateInAnyOrderAssertion` as well as the override for `searchAndCreateAssertion` here.